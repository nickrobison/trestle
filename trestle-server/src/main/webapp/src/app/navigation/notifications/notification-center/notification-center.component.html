<div *ngFor="let notification of notifications | async">
  <app-toast [notification]="notification"></app-toast>
</div>

