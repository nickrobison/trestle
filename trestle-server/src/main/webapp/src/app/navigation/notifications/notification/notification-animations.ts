import {animate, AnimationTriggerMetadata, state, style, transition, trigger} from "@angular/animations";

export const notificationAnimations: {
  readonly fadeNotification: AnimationTriggerMetadata,
  readonly buttonNotification: AnimationTriggerMetadata,
} = {
  fadeNotification: trigger('fadeAnimation', [
    state('in', style({opacity: 1})),
    transition('void => *', [style({
      opacity: 0.5
    }),
      animate('{{ fadeIn }}ms')]),
    transition('default => closing', animate('{{ fadeOut }}ms', style({ opacity: 0 })),
      ),
  ]),
  buttonNotification: trigger('buttonAnimation', [
    state('out', style({opacity: 0.5})),
    transition('in => out', [style({
      opacity: 0.5
    }),
    animate('500ms')]),
    transition('out => in', [style({
        opacity: 1
      }),
      animate('500ms')]),
  ]),
};

export type NotificationAnimationState = 'default' | 'closing';

export type NotificationButtonState = 'in' | 'out';
