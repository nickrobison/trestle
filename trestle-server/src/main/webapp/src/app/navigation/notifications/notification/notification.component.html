<div class="toast"
     [@fadeAnimation]="{value: animationState, params: {fadeIn: 1000, fadeOut: 1000}}"
     (@fadeAnimation.done)="onFadeFinished($event)"
>
  <div class="timer" [attr.data-status]="notification.state" style="{{'--duration: ' + notificationLifetime + ';'}}">
    <div class="background"></div>
    <div class="timebar"></div>
  </div>
  <div class="msg-text">
    {{getMessage()}}
  </div>
  <mat-icon (click)="triggerClose()"
            (mouseenter)="this.buttonState = 'in'"
            (mouseleave)="this.buttonState = 'out'"
            [@buttonAnimation]="{value: buttonState}">close
  </mat-icon>
</div>

