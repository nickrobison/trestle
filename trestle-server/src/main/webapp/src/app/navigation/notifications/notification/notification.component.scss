.toast {
  position: relative;
  display: flex;
  justify-content: space-around;
  margin: 0 10px 20px 0;
  padding: 10px 15px 10px 10px;
  width: 290px;
  background: #fff;
  border-width: 1px;
  border-style: solid;
  border-color: #dddddd #d6d6d6 #cfcfcf;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.11);
}

.msg-text {
  margin-left: 5px;
}

$bar-width: 5px;
$opacity: 80%;

.timer .background {
  height: 100%;
  width: $bar-width;
  background: yellowgreen;
  position: absolute;
  left: 0;
  bottom: 0;
  opacity: $opacity;
}

.timer .timebar {
  height: 100%;
  width: $bar-width;
  background: green;
  position: absolute;
  left: 0;
  bottom: 0;
  opacity: $opacity;
  animation: roundtime calc(var(--duration) * 1s) linear forwards;
  transform-origin: bottom;
}

.timer[data-status="notification"] .background {
  background: yellowgreen;
}

.timer[data-status="notification"] .timebar {
  background: green;
}

.timer[data-status="error"] .background {
  background: indianred;
}

.timer[data-status="error"] .timebar {
  background: red;
}

@keyframes roundtime {
  to {
    transform: scaleY(0);
  }
}
