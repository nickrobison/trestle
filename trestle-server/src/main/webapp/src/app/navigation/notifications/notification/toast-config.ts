export class ToastData {
  msg: string;
}
