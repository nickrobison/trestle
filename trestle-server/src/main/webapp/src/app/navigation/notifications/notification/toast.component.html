<div class="toast">
  <p>toast works!</p>
</div>

