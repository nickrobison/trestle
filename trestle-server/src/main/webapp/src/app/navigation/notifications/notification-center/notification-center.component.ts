import {Component, OnInit} from '@angular/core';
import {selectNotificationsFromNotifications, TrestleNotification} from "../../../reducers/notification.reducers";
import {Observable} from "rxjs";
import {select, Store} from "@ngrx/store";
import {State} from "../../../reducers";

@Component({
  selector: 'app-notification-center',
  templateUrl: './notification-center.component.html',
  styleUrls: ['./notification-center.component.scss']
})
export class NotificationCenterComponent implements OnInit {

  public notifications: Observable<TrestleNotification[]>;

  constructor(private store: Store<State>) {
    // Not used
  }

  ngOnInit(): void {
    this.notifications = this.store
      .pipe(select(selectNotificationsFromNotifications));
  }
}
