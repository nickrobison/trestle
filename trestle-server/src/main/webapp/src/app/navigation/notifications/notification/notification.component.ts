import {Component, Input, OnDestroy, OnInit} from '@angular/core';
import {TrestleNotification} from "../../../reducers/notification.reducers";
import {Store} from "@ngrx/store";
import {State} from "../../../reducers";
import {removeNotification} from "../../../actions/notification.actions";
import {notificationAnimations, NotificationAnimationState, NotificationButtonState} from "./notification-animations";
import {AnimationEvent} from "@angular/animations";

@Component({
  selector: 'app-toast',
  templateUrl: './notification.component.html',
  styleUrls: ['./notification.component.scss'],
  animations: [notificationAnimations.fadeNotification, notificationAnimations.buttonNotification],
})
export class NotificationComponent implements OnInit, OnDestroy {

  @Input()
  notification: TrestleNotification;
  readonly notificationLifetime = 5;
  animationState: NotificationAnimationState = 'default';
  buttonState: NotificationButtonState = 'out';
  private intervalId: number;

  constructor(private store: Store<State>) {
  }

  ngOnInit(): void {
    this.intervalId = setTimeout(() => this.triggerClose(), this.notificationLifetime * 1000);
  }

  ngOnDestroy() {
    clearTimeout(this.intervalId);
  }

  onFadeFinished(event: AnimationEvent) {
    const {toState} = event;
    const isFadeOut = (toState as NotificationAnimationState) === 'closing';
    const itFinished = this.animationState === 'closing';
    if (itFinished && isFadeOut) {
      this.close();
    }
  }

  triggerClose() {
    this.animationState = 'closing';
  }

  private close() {
    this.store.dispatch(removeNotification({notification: this.notification}));
  }

  getMessage(): string {
    switch (this.notification.state) {
      case 'error':
        return this.notification.error.message;
      default:
        return this.notification.msg;
    }
  }

}
