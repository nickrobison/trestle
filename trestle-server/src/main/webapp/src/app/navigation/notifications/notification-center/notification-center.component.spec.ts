import {ComponentFixture, TestBed, waitForAsync} from '@angular/core/testing';

import {NotificationCenterComponent} from './notification-center.component';
import {NotificationComponent} from '../notification/notification.component';
import {MaterialModule} from '../../../material/material.module';
import {MockStore, provideMockStore} from '@ngrx/store/testing';
import {MemoizedSelector} from '@ngrx/store';
import {State} from '../../../reducers';
import {
  selectNotificationsFromNotifications,
  TrestleError,
  TrestleMessage,
  TrestleNotification
} from '../../../reducers/notification.reducers';

describe('NotificationCenterComponent', () => {
  let component: NotificationCenterComponent;
  let fixture: ComponentFixture<NotificationCenterComponent>;
  let mockStore: MockStore;
  let mockNotificationSelector: MemoizedSelector<State, TrestleNotification[]>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ NotificationCenterComponent, NotificationComponent ],
      providers: [provideMockStore()],
      imports: [MaterialModule],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationCenterComponent);
    component = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);
    mockNotificationSelector = mockStore.overrideSelector(
      selectNotificationsFromNotifications,
      []
    );
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should have notifications', () => {
    const tError: TrestleError = {
      state: 'error',
      error: new Error('Cannot search for individuals'),
    };
    const tMessage: TrestleMessage = {
      state: 'notification',
      msg: 'Hello, Im a message'
    };
    const expected = [tError, tMessage];
    mockNotificationSelector.setResult(expected);
    fixture.detectChanges();
    component.notifications.subscribe(notifications => {
      expect(notifications).toEqual(expected);
    });
    expect(fixture).toMatchSnapshot();
  });
});
