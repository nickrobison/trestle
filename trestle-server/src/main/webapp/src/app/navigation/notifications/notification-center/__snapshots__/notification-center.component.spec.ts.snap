// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationCenterComponent should create 1`] = `
<app-notification-center
  notifications={[Function Store]}
  store={[Function MockStore]}
>
  
</app-notification-center>
`;

exports[`NotificationCenterComponent should have notifications 1`] = `
<app-notification-center
  notifications={[Function Store]}
  store={[Function MockStore]}
>
  
</app-notification-center>
`;
